jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp

```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+  bool is_enter,
+  const exprt &object)
+{
+  code_typet type;
+  type.return_type() = void_typet();
+  type.parameters().resize(1);
+  type.parameters()[0].type() = java_reference_type(void_typet());
+  std::string symbol;
+
+  if(is_enter)
+    symbol="java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V";
+  else
+    symbol="java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V";
+
+  symbol_tablet::symbolst::const_iterator it
+    = symbol_table.symbols.find(symbol);
+
+  // If the functions for monitorenter or monitorexit do not exist
+  // (this can only happen if the java-models libary was not downloaded).
+  // We implement them as skips because symex would crash, since it cannot find
+  // the function in the symbol table
+  if(it == symbol_table.symbols.end())
+    return code_skipt();
+
+  // Otherwise we create a function call
+  code_function_callt call;
+  call.function() = symbol_exprt(symbol, type);
+  call.lhs().make_nil();
+  call.arguments().push_back(object);
+  return call;
+}
+
+/// Introduces a monitorexit before every return recursively
+/// \param code current element to check
+/// \param monitorexit codet to insert before the return.
+static void monitor_exits(
+  codet &code,
+  const codet &monitorexit)
+{
+  const irep_idt &statement=code.get_statement();
+  if(statement == ID_return)
+  {
+    // Relace the return with a monitor exit plus return block
+    code_blockt return_block;
+    return_block.add(monitorexit);
+    return_block.move_to_operands(code);
+    code = return_block;
+  }
+  else if((statement == ID_label)
+    || (statement == ID_block)
+    || (statement == ID_decl_block))
+  {
+    // If label or block found, explore the code inside the block
+    Forall_operands(it, code)
+    {
+      codet &sub_code = to_code(*it);
+      monitor_exits(sub_code, monitorexit);
+    }
+  }
+}
+
+/// Transforms the codet stored in in the value field of \p symbol. This
+/// expected to be a codet that needs to be synchronized, the transformation
+/// makes it thread-safe as described in the documentation of function
+/// convert_synchronized_methods
+///
+/// The resulting thread-safe codet is stored in the value field of \p symbol.
+///
+/// \param symbol_table: a symbol_table
+/// \param symbol: writeable symbol hosting code to synchronized
+/// \param sync_object: object to use as a lock
+static void instrument_synchronized_code(
+  symbol_tablet &symbol_table,
+  symbolt &symbol,
+  const exprt &sync_object)
+{
+    codet &code = to_code(symbol.value);
+
+    // Get interposition code for the monitor lock/unlock
+    codet monitorenter = get_monitor_call(symbol_table, true, sync_object);
+    codet monitorexit = get_monitor_call(symbol_table, false, sync_object);
+
+    // Create a unique catch label and empty throw type (i.e any)
+    // and catch-push them at the beginning of the code (i.e begin try)
+    code_push_catcht catch_push;
+    irep_idt handler("pc_synchronized_catch");
```

Comment:
Is this a magic string?

Reply:
Yes, it  is used to create a unique catch label. It only used locally in that function right now,  should it be prefixed with 'CPROVER'?

Reply:
How would you guarantee that it genuinely is unique? I'd recommend to use some string that cannot possibly be used by a user, for example by using a `-` somewhere in there.

Reply:
done.